From e2aaca4678abad93c0fab889fc2d26d8d764b0af Mon Sep 17 00:00:00 2001
From: Gustavo Diaz Prado <x0083741@ti.com>
Date: Sat, 5 Mar 2011 14:26:12 -0600
Subject: omap4: omapfb: Align framebuffer stride to 8 pixels

The SGX GPU requires the pixel stride to be aligned to 8 pixels
to prevent any chance of corruption on a render target. This
patch ensures the framebuffer complies with this hardware
alignment and ensures enough memory is allocated properly with
the new stride.

Change-Id: I13ba8a1fe372f87e6dc2a40770fe9e46304efe5a
Signed-off-by: Gustavo Diaz Prado <x0083741@ti.com>
---
 drivers/video/omap2/omapfb/omapfb-main.c |   10 +++++++++-
 1 files changed, 9 insertions(+), 1 deletions(-)

diff --git a/drivers/video/omap2/omapfb/omapfb-main.c b/drivers/video/omap2/omapfb/omapfb-main.c
index 0ed1b12..50eae05 100644
--- a/drivers/video/omap2/omapfb/omapfb-main.c
+++ b/drivers/video/omap2/omapfb/omapfb-main.c
@@ -45,6 +45,10 @@
 #define OMAPFB_PLANE_XRES_MIN		8
 #define OMAPFB_PLANE_YRES_MIN		8
 
+#ifdef CONFIG_SGX540
+#define SGX_HW_ALIGN_PIXELS		8
+#endif
+
 static char *def_mode;
 static char *def_vram;
 static int def_vrfb;
@@ -657,8 +661,12 @@ void set_fb_fix(struct fb_info *fbi)
 
 		fix->smem_len = var->yres_virtual * fix->line_length;
 	} else if (ofbi->rotation_type != OMAP_DSS_ROT_TILER) {
+		int aligned_width = var->xres_virtual;
+#ifdef CONFIG_SGX540
+		aligned_width = ALIGN(aligned_width, SGX_HW_ALIGN_PIXELS);
+#endif
 		fix->line_length =
-			(var->xres_virtual * var->bits_per_pixel) >> 3;
+			(aligned_width * var->bits_per_pixel) >> 3;
 
 		/* tiler line length is set during allocation, and cannot
 		   be changed */
-- 
1.6.0.4

